### PR TITLE
Add timestamp functions to Debug::print.

### PR DESCRIPTION
**Pull request by jamiefaye**
Code that prepends a 32 bit timestamp to Debug::print messages. Also includes a RoutineTimer that works like a stopwatch and includes durations.

### CODE REVIEW

File: src/deluge/drivers/uart/uart.h [hunk 1]
```diff
@@ -32,7 +32,7 @@
  * and set HAVE_RTT to 0. Though this configuration hasn’t been tested for a while...
  */
 #ifndef ENABLE_TEXT_OUTPUT
-#define ENABLE_TEXT_OUTPUT 0
+#define ENABLE_TEXT_OUTPUT 1
```
Comment (m-m-adams) [inline src/deluge/drivers/uart/uart.h:35]: Is this intentional or did it slip in? The cmake config assumes the default is 0 and sets it to 1 in debug and relwithdebinfo builds

File: src/deluge/processing/engines/audio_engine.cpp [hunk 2]
```diff
@@ -247,7 +247,7 @@ Voice* cullVoice(bool saveVoice, bool justDoFastRelease) {
 
 	uint32_t bestRating = 0;
 	Voice* bestVoice = NULL;
-
+	Debug::println("cull");
```
Comment (m-m-adams) [inline src/deluge/processing/engines/audio_engine.cpp:250]: Doesn't this lead to duplicated lines? There's printing in the routine function for culls as well

Comment (jamiefaye) [reply src/deluge/processing/engines/audio_engine.cpp:250]: It is a duplicate that won't go in the final.

File: src/deluge/drivers/uart/uart.h (hunk 1 above)
Comment (jamiefaye) [reply src/deluge/drivers/uart/uart.h:35]: We can drop this part too and let cmake handle matters.

**Review by jamiefaye — COMMENTED** [review]
Don't be afraid to tear into this with your comments.

File: src/deluge/io/debug/print.cpp [hunk 3]
```diff
@@ -27,11 +27,109 @@ extern "C" {
 
 namespace Debug {
 
+constexpr uint32_t kNumSamplesBetweenReports = 44100;
+
+constexpr bool SysExLoggingEnabled = true;
```
Comment (m-m-adams) [inline src/deluge/io/debug/print.cpp:32]: can this be a pre processor define? It'll make it easier to configure

File: src/deluge/io/debug/print.cpp [hunk 4]
```diff
@@ -67,4 +166,306 @@ void print(int32_t number) {
...
+    : label(label), timeBase(timeBase), active(false), accumulator(0), count(0), t0(0) {
+}
+
+void AverageVOT::note(uint32_t value) {
+#if ENABLE_TEXT_OUTPUT
+	if (!active) {
+		count = 0;
+		accumulator = 0;
+		active = true;
+		asm volatile("MRC p15, 0, %0, c9, c13, 0" : "=r"(t0) :);
```
[the remark this answers is not in the document]
Comment (jamiefaye) [reply src/deluge/io/debug/print.cpp:441]: It already has one! I should be using that.

File: src/deluge/io/debug/print.cpp (hunk 3 above)
Comment (jamiefaye) [reply src/deluge/io/debug/print.cpp]: I did add a #define SYSEX_LOGGING_ENABLED which is defaulted to false if not defined by the build configuration

File: src/deluge/io/debug/print.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jamiefaye) [reply src/deluge/io/debug/print.cpp]: I added an "always inlined" function that uses a reference to the variable to be updated.

File: src/deluge/io/debug/print.cpp (hunk 3 above)
Comment (m-m-adams) [reply src/deluge/io/debug/print.cpp]: Thanks! I just want to make it easy to find if it needs to be turned on/off

**Review by m-m-adams — APPROVED** [review]

**Pull request merged**